x/leverage/keeper/accounts_summary.go

```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		return nil, err
+	}
+	lastBorrowedValue, err := q.VisibleTokenValue(ctx, borrowed, types.PriceModeQuery)
+	if err != nil {
+		return nil, err
+	}
+	lastCollateralValue, err := q.VisibleCollateralValue(ctx, collateral, types.PriceModeQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// these use leverage-like prices: the lower of spot or historic price for supplied tokens and higher for borrowed.
+	// unlike transactions, this query will use expired prices instead of skipping them.
+	suppliedValue, err := q.VisibleTokenValue(ctx, supplied, types.PriceModeQueryLow)
+	if err != nil {
+		return nil, err
+	}
+	collateralValue, err := q.VisibleCollateralValue(ctx, collateral, types.PriceModeQueryLow)
+	if err != nil {
+		return nil, err
+	}
+	borrowedValue, err := q.VisibleTokenValue(ctx, borrowed, types.PriceModeQueryHigh)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &types.QueryAccountSummaryResponse{
+		SuppliedValue:       suppliedValue,
+		CollateralValue:     collateralValue,
+		BorrowedValue:       borrowedValue,
+		SpotSuppliedValue:   lastSuppliedValue,
+		SpotCollateralValue: lastCollateralValue,
+		SpotBorrowedValue:   lastBorrowedValue,
+	}
+
+	// values computed from position use the same prices found in leverage logic:
+	// using the lower of spot or historic prices for each collateral token
+	// and the higher of spot or historic prices for each borrowed token
+	// skips collateral tokens with missing prices, but errors on borrow tokens missing prices
+	// (for oracle errors only the relevant response fields will be left nil)
+	ap, err := q.GetAccountPosition(ctx, addr, false)
+	if nonOracleError(err) {
+		return nil, err
+	}
+	if err == nil {
+		// on missing borrow price, borrow limit is nil
+		borrowLimit := ap.Limit()
+		resp.BorrowLimit = &borrowLimit
+	}
+
+	// liquidation threshold shown here as it is used in leverage logic: using spot prices.
+	// skips borrowed tokens with missing prices, but errors on collateral missing prices
+	// (for oracle errors only the relevant response fields will be left nil)
+	ap, err = q.GetAccountPosition(ctx, addr, true)
+	if nonOracleError(err) {
+		return nil, err
+	}
+	if err == nil {
+		// on missing collateral price, liquidation threshold is nil
+		liquidationThreshold := ap.Limit()
+		resp.LiquidationThreshold = &liquidationThreshold
+	}
+
+	return resp, nil
+}
+
+// AccountsSummary implements types.QueryServer.
+func (q Querier) AccountsSummary(goCtx context.Context, req *types.QueryAccountsSummary) (
+	*types.QueryAccountsSummaryResponse, error,
+) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	// get the all accounts
+	store := ctx.KVStore(q.akStoreKey)
+	accountsStore := prefix.NewStore(store, authtypes.AddressStoreKeyPrefix)
+
+	var accounts []*types.AccountSummary
+	pageRes, err := query.Paginate(accountsStore, req.Pagination, func(key, value []byte) error {
```

Comment:
You should iterate over supply KV prefix, no need to iterate over all supplies

Reply:
With bank supply KV prefix, we only get the denom + total supply of denom , but we need accounts with balances